### ISSUE TITLE
[SUGGESTION] Simplify the Parser

### ISSUE THREAD

**Issue by tdeeb** [teamradish/TRTwitchPlaysBot#48]
**Describe your suggestion**
TRBot's parser was originally adapted from TwitchPlays_Everything's parser, and since then it has been enhanced with greater performance and a slightly more robust syntax. However, it's problems are still there: it's by-and-large monolithic and tries to account for every possible input on every platform.

This is not ideal for many input methods, such as touch screens, motion controls, and mouse movements. For example, it's not clear what "left500ms" would mean on a mouse - this raises questions, such as what the speed of the mouse is and whether 1 millisecond equals 1 pixel.

As a result, TRBot needs to be able to support a different syntax for different platforms so playing with it can make sense. This can also open up more configuration opportunities, such as removing holds, releases, multi-controller inputs, or changing duration syntax (Ex. "up0.5s"). More details TBD.

**Additional context**
[tpp-core](https://github.com/TwitchPlaysPokemon/tpp-core/tree/master/Inputting) has a similar system to what we're looking for, but their inputs are separated with commas, whereas TRBot has no clear separation. We'll need to come up with something that can handle all the intricacies specific to TRBot, such as multi-controller inputs, macros, and so on.

**Comment by tdeeb** [comment]
Notes:
- tpp-core adds numbered labels to each input [in the regex](https://github.com/TwitchPlaysPokemon/tpp-core/blob/master/Inputting/Parsing/BareInputParser.cs#L39). This makes it much easier to parse since it's possible to simply find groups with given names. This extends to holds and repetitions.
- Our parser, like tpp-core's, should eventually be easily testable. Right now parsing inputs is very complicated and requires a lot of pre-processing. We should look to trim this down as much as possible.

**Comment by tdeeb** [comment]
More thoughts:

I really like tpp-core's concept of [InputDefinitions](https://github.com/TwitchPlaysPokemon/tpp-core/blob/master/Inputting/IInputDefinition.cs).

I think we can break down the idea InputDefinitions even further for TRBot via Parser Components. Parser Components would be the bits that involve parsing each input, including:
  - The syntax for controller inputs: `(?:<port>&1)?`
  - The syntax for holds: `(?:<hold>_)?`
  - The syntax for the input itself: `(?:<input>(start|left|b))`
  
If done right, this should allow for full flexibility in the parser's syntax.

Regarding the inputs themselves, we will have to change them since there's no way to accomodate for every type of input as-is. Surface level, I'm thinking of having a dictionary for the parts of the input, but I'm not yet sure if that's ideal.

**Comment by tdeeb** [comment]
This will have to be pushed back due to its complexity.

While we can change the way inputs are parsed, executing them is a different story. We would need to modify the virtual controllers to be able to execute different types of inputs, such as a touchpad or mouse, which is inflexible. Every virtual controller would need to define methods for pressing and releasing these types of inputs.

Furthermore, the InputHandler would need to be refactored as well, and we would need different types of input definitions (Touch, Mouse, etc) that are accounted for in the console.

In short, everything needs to be changed, and there's no solid plan for this yet. It may be more suitable for a 3.0 release.

However, I'm all for simplifying the parser, so we should at least go through with that.

**Comment by tdeeb** [comment]
The parser has been simplified for the 2.2 release. Going forwards, `Parser` is deprecated in favor of the `IParser` interface.
